src/java/org/apache/cassandra/index/sai/disk/v1/kdtree/BKDTreeRamBuffer.java

```diff
@@ -0,0 +1,186 @@
+        this.pointNumBytes = pointNumBytes;
+
+        this.bytes = new ByteBlockPool(new ByteBlockPool.DirectTrackingAllocator(bytesUsed));
+
+        packedValue = new byte[pointNumBytes];
+        packedBytesLength = pointNumBytes;
+
+        docIDsBuilder = PackedLongValues.deltaPackedBuilder(PackedInts.COMPACT);
+        bytesUsed.addAndGet(docIDsBuilder.ramBytesUsed());
+    }
+
+    @Override
+    public long ramBytesUsed()
+    {
+        return bytesUsed.get();
+    }
+
+    public int numRows()
+    {
+        return numRows;
+    }
+
+    public long addPackedValue(int segmentRowId, BytesRef value)
+    {
+        ensureOpen();
+        
+        if (value.length != packedBytesLength)
+        {
+            throw new IllegalArgumentException("The value has length=" + value.length + " but should be " + pointNumBytes);
+        }
+
+        long startingBytesUsed = bytesUsed.get();
+        long startingDocIDsBytesUsed = docIDsBuilder.ramBytesUsed();
+
+        docIDsBuilder.add(segmentRowId);
+        bytes.append(value);
+
+        if (segmentRowId != lastSegmentRowID)
+        {
+            numRows++;
+            lastSegmentRowID = segmentRowId;
+        }
+
+        numPoints++;
+
+        long docIDsAllocatedBytes = docIDsBuilder.ramBytesUsed() - startingDocIDsBytesUsed;
+        long endingBytesAllocated = bytesUsed.addAndGet(docIDsAllocatedBytes);
+        
+        return endingBytesAllocated - startingBytesUsed;
+    }
+
+    public MutableOneDimPointValues asPointValues()
+    {
+        ensureOpen();
+        // building packed longs is destructive
+        closed = true;
+        final PackedLongValues docIDs = docIDsBuilder.build();
+        return new MutableOneDimPointValues()
+        {
+            final int[] ords = new int[numPoints];
+
+            {
+                for (int i = 0; i < numPoints; ++i)
+                {
+                    ords[i] = i;
+                }
+            }
+
+            @Override
+            public void getValue(int i, BytesRef packedValue)
+            {
+                final long offset = (long) packedBytesLength * (long) ords[i];
+                packedValue.length = packedBytesLength;
+                bytes.setRawBytesRef(packedValue, offset);
+            }
+
+            @Override
+            public byte getByteAt(int i, int k)
+            {
+                final long offset = (long) packedBytesLength * (long) ords[i] + (long) k;
+
+                return bytes.readByte(offset);
+            }
+
+            @Override
+            public int getDocID(int i)
+            {
+                return Math.toIntExact(docIDs.get(ords[i]));
+            }
+
+            @Override
+            public void swap(int i, int j)
+            {
+                int tmp = ords[i];
+                ords[i] = ords[j];
+                ords[j] = tmp;
+            }
+
+            @Override
+            public void intersect(IntersectVisitor visitor) throws IOException
```

Comment:
nit: Should `intersect()` be `visit()` or `visitAll()` or something along those lines?

Reply:
I have been debating this one and came to the conclusion that `intersect` is probably the correct name. Although, in the majority of cases, the implementation is just an iteration of values, it is also used to test values before passing them to the visitor.

I don't feel that `visit` is correct because this isn't the visitor. It could be `visitAll` but, as above, this is not true in all cases.

Reply:
It have thought more about this and have decided that we leave this as `intersect` for now but with [CASSANDRA-18598](https://issues.apache.org/jira/browse/CASSANDRA-18598) we will have no dependency on Lucene `PointValues` and can remove the `intersection` / `visitor` patterns completely in favour of making the reader `Iterable`.